File: languages/go/src/operations/quickSort.go
package operations

import (
	"github.com/benchmark-playground/go/utils"
)

func partition[T utils.Comparable[T]](arr []T) int {
	len := len(arr)
	pivot := len - 1
	storeIndex := 0

	for i := range len {
		if arr[i].CompareTo(arr[pivot]) < 0 {
			arr[i], arr[storeIndex] = arr[storeIndex], arr[i]
			storeIndex++
		}
	}
	arr[storeIndex], arr[pivot] = arr[pivot], arr[storeIndex]
	return storeIndex
}

func QuickSort[T utils.Comparable[T]](arr []T) []T {
	if len(arr) > 1 {
		pivot := partition(arr)
		QuickSort(arr[:pivot])
		QuickSort(arr[pivot+1:])
	}
	return arr
}

File: languages/c++/src/operations/quick_sort.cpp
#include <cstring>

#include "operations.h"

namespace operations {
    inline void swap(char* arr[], int index1, int index2) {
        char* temp = arr[index1];
        arr[index1] = arr[index2];
        arr[index2] = temp;
    }

    // Partition function to find the pivot position
    inline int partition(char *arr[], int len) {
        char *pivot = arr[len - 1]; // Choose the last element as pivot
        int storeIndex = 0;

        for (int i = 0; i < len - 1; i++) {
            if (strcmp(arr[i], pivot) < 0) {    // Compare strings
                swap(arr, i, storeIndex);
                storeIndex++;
            }
        }
        swap(arr, storeIndex, len - 1);
        return storeIndex;
    }

    char** quickSort(char *arr[], int len) {
        if (len > 1) {
            int pivot = partition(arr, len);
            quickSort(arr, pivot);             // Recursively sort the left part
            quickSort(arr + pivot + 1, len - pivot - 1); // Recursively sort the right part
        }
        return arr;
    }
}